src/workerd/api/crypto/rsa.c++

```diff
@@ -225,6 +253,196 @@ SubtleCrypto::JsonWebKey Rsa::toJwk(KeyType keyType,
+                                 "Second Prime Factor parameter (\"q\").");
+    auto dp = JSG_REQUIRE_NONNULL(jwk.dp.map([](auto& str) { return str.asPtr(); }), Error,
+                                  "Invalid RSA key in JSON Web Key; missing or invalid "
+                                  "First Factor CRT Exponent parameter (\"dp\").");
+    auto dq = JSG_REQUIRE_NONNULL(jwk.dq.map([](auto& str) { return str.asPtr(); }), Error,
+                                  "Invalid RSA key in JSON Web Key; missing or invalid "
+                                  "Second Factor CRT Exponent parameter (\"dq\").");
+    auto qi = JSG_REQUIRE_NONNULL(jwk.qi.map([](auto& str) { return str.asPtr(); }), Error,
+                                  "Invalid RSA key in JSON Web Key; missing or invalid "
+                                  "First CRT Coefficient parameter (\"qi\").");
+    auto dDecoded = toBignumUnowned(simdutfBase64UrlDecodeChecked(d,
+        "Invalid RSA key in JSON Web Key"_kj));
+    auto pDecoded = toBignumUnowned(simdutfBase64UrlDecodeChecked(p, kInvalidBase64Error));
+    auto qDecoded = toBignumUnowned(simdutfBase64UrlDecodeChecked(q, kInvalidBase64Error));
+    auto dpDecoded = toBignumUnowned(simdutfBase64UrlDecodeChecked(dp, kInvalidBase64Error));
+    auto dqDecoded = toBignumUnowned(simdutfBase64UrlDecodeChecked(dq, kInvalidBase64Error));
+    auto qiDecoded = toBignumUnowned(simdutfBase64UrlDecodeChecked(qi, kInvalidBase64Error));
+
+    JSG_REQUIRE(RSA_set0_key(rsa.get(), nullptr, nullptr, dDecoded) == 1, Error,
+                "Invalid RSA key in JSON Web Key; failed to set private exponent");
+    JSG_REQUIRE(RSA_set0_factors(rsa.get(), pDecoded, qDecoded) == 1, Error,
+                "Invalid RSA key in JSON Web Key; failed to set prime factors");
+    JSG_REQUIRE(RSA_set0_crt_params(rsa.get(), dpDecoded, dqDecoded, qiDecoded) == 1, Error,
+                "Invalid RSA key in JSON Web Key; failed to set CRT parameters");
+  }
+
+  auto evpPkey = OSSL_NEW(EVP_PKEY);
+  KJ_ASSERT(EVP_PKEY_set1_RSA(evpPkey.get(), rsa.get()) == 1);
+
+  auto usages = keyType == KeyType::PRIVATE ? CryptoKeyUsageSet::privateKeyMask()
+                                            : CryptoKeyUsageSet::publicKeyMask();
+  return AsymmetricKeyData {
+    kj::mv(evpPkey),
+    keyType,
+    usages
+  };
+}
+
+kj::String Rsa::toPem(KeyEncoding encoding,
+                      KeyType keyType,
+                      kj::Maybe<CipherOptions> options) const {
+  ClearErrorOnReturn clearErrorOnReturn;
+  auto bio = OSSL_BIO_MEM();
+  switch (keyType) {
+    case KeyType::PUBLIC: {
+      switch (encoding) {
+        case KeyEncoding::PKCS1: {
+          JSG_REQUIRE(PEM_write_bio_RSAPublicKey(bio.get(), rsa) == 1, Error,
+              "Failed to write RSA public key to PEM", tryDescribeOpensslErrors());
+          break;
+        }
+        case workerd::api::KeyEncoding::SPKI: {
+          JSG_REQUIRE(PEM_write_bio_RSA_PUBKEY(bio.get(), rsa) == 1, Error,
```

Comment:
Off-topic: It's really confusing that `PEM_write_bio_RSAPublicKey`/`PEM_write_bio_RSA_PUBKEY` have nearly the same name but are used for different data formats...

Reply:
Yep... have had this same conversation before in node ;-)